java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java

```diff
@@ -0,0 +1,415 @@
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(id);
+    }
+
+    @Override
+    public String toString() {
+      return "SV#" + id;
+    }
+
+    @Override
+    public void computedFrom(List<SymbolicValue> symbolicValues) {
+      // no op in general case
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, NullConstraint nullConstraint) {
+      Object data = programState.constraints.get(this);
+      if (data instanceof NullConstraint) {
+        NullConstraint nc = (NullConstraint) data;
+        if ((NullConstraint.NULL.equals(nullConstraint) && NullConstraint.NOT_NULL.equals(nc)) ||
+          (NullConstraint.NULL.equals(nc) && NullConstraint.NOT_NULL.equals(nullConstraint))) {
+          // setting null where value is known to be non null or the contrary
+          return ImmutableList.of();
+        }
+      }
+      if (data == null || !data.equals(nullConstraint)) {
+        Map<SymbolicValue, Object> temp = Maps.newHashMap(programState.constraints);
+        temp.put(this, nullConstraint);
+        return ImmutableList.of(new ProgramState(programState.values, temp, programState.visitedPoints, programState.stack));
+      }
+      return ImmutableList.of(programState);
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
+      Object data = programState.constraints.get(this);
+      // update program state only for a different constraint
+      if (data instanceof BooleanConstraint) {
+        BooleanConstraint bc = (BooleanConstraint) data;
+        if ((BooleanConstraint.TRUE.equals(booleanConstraint) && BooleanConstraint.FALSE.equals(bc)) ||
+          (BooleanConstraint.TRUE.equals(bc) && BooleanConstraint.FALSE.equals(booleanConstraint))) {
+          // setting null where value is known to be non null or the contrary
+          return ImmutableList.of();
+        }
+      }
+      if (data == null || !data.equals(booleanConstraint)) {
+
+        // store constraint only if symbolic value can be reached by a symbol.
+        if (programState.values.containsValue(this)) {
+          Map<SymbolicValue, Object> temp = Maps.newHashMap(programState.constraints);
+          temp.put(this, booleanConstraint);
+          return ImmutableList.of(new ProgramState(programState.values, temp, programState.visitedPoints, programState.stack));
+        }
+      }
+      return ImmutableList.of(programState);
+    }
+
+    @Override
+    public ProgramState setSingleConstraint(ProgramState programState, NullConstraint nullConstraint) {
+      final List<ProgramState> states = setConstraint(programState, nullConstraint);
+      if (states.size() != 1) {
+        throw new IllegalStateException("Only a single program state is expected at this location");
+      }
+      return states.get(0);
+    }
+  }
+
+  abstract class BinarySymbolicValue extends ObjectSymbolicValue {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Override this superclass' "equals" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2160)